### ISSUE TITLE
Add ability to cache Modules in memory

### ISSUE THREAD

**Issue by evgenykuzyakov** [wasmerio/wasmer#174]
Hi,

I've tried to cache Module completely in memory, but most of the APIs are crate specific. I was able to clone Module and reuse it for the next instance. But it seems crashing my tests.

One option would be just add Clone on Cache.
And if we have Clone on Cache, it would make sense to remove unsafe from `into_module` for instances that were cloned and not recovered from disk

**Comment by syrusakbary** [comment]
Hi @evgenykuzyakov,

We are working on refactoring a bit the Cache mechanism to make it easier to use and extend.

Here's an example on how it could be used:
https://gist.github.com/syrusakbary/fde50e5631458af5f8479ff8999a670d

We'd love to hear your feedback! :)

**Comment by evgenykuzyakov** [comment]
That looks good. We've currently switched to in-memory cache using temp branch with Clone. One thing I'm also doing is to preprocess the original WASM code before compiling it. So the hash of the cached code is different from the cache's key. It should be fine looking at the code above.

**Comment by lachlansneff** [comment]
@evgenykuzyakov Just to specify, you're running into crashes when reusing a `Module`?

**Comment by evgenykuzyakov** [comment]
@lachlansneff Yes. I didn't debug it a lot. I was using `cached!` and was cloning `Arc<Module>` within it. So it was potentially using the same `Module` in memory to instantiate different instances. It was crashing `cargo test` without panic in random places (sometimes more tests, sometimes less).
So I decided to work with `Cache` instead after this.

**Comment by lachlansneff** [comment]
Hmm, that's odd. A `Module` is designed to be reusable. I'll look into it.

**Comment by lachlansneff** [comment]
I've been trying to reproduce the module reuse crash and have been unable to.

Here's the code I've been testing with:

```rust
static WASM: &'static [u8] = &[
   0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00, 0x01, 0x06, 0x01, 0x60,
   0x01, 0x7f, 0x01, 0x7f, 0x03, 0x02, 0x01, 0x00, 0x07, 0x0b, 0x01, 0x07,
   0x61, 0x64, 0x64, 0x5f, 0x6f, 0x6e, 0x65, 0x00, 0x00, 0x0a, 0x09, 0x01,
   0x07, 0x00, 0x20, 0x00, 0x41, 0x01, 0x6a, 0x0b, 0x00, 0x1a, 0x04, 0x6e,
   0x61, 0x6d, 0x65, 0x01, 0x0a, 0x01, 0x00, 0x07, 0x61, 0x64, 0x64, 0x5f,
   0x6f, 0x6e, 0x65, 0x02, 0x07, 0x01, 0x00, 0x01, 0x00, 0x02, 0x70, 0x30,
];

use wasmer_runtime::{
   compile,
   imports,
   error,
   Func,
};

fn main() -> error::Result<()> {
    // We're not importing anything, so make an empty import object.
    let import_object = imports! {};

    let module = compile(WASM)?;

    for _ in 0..100000 {
        let instance = module.instantiate(&import_object)?;

        let add_one: Func<i32, i32> = instance.func("add_one")?;

        let value = add_one.call(42)?;

        assert_eq!(value, 43);
    }

    Ok(())
}
```

**Comment by syrusakbary** [comment]
Quick update on the caching issue.

We just merged #198 and #200.
Thanks to those, the API for creating a custom in-memory cache should be quite simple.

Here's an example on how to create a Cache implementation (`FileSystemCache`):
https://github.com/wasmerio/wasmer/blob/master/lib/runtime/src/cache.rs#L80-L108

And here's how to use it (should be the same for other `Cache` implementations):
https://github.com/wasmerio/wasmer/blob/master/src/bin/wasmer.rs#L114-L139

Let us know if you have any questions!
